### ISSUE TITLE
Fail to hotplug device with `iommu=on`

### ISSUE THREAD

**Issue by likebreath** [cloud-hypervisor/cloud-hypervisor#3205]
If we start a guest that does not require virtual `IOMMU`, we will hit the following error if we hotplug a device with `iommu=on` after VM boot:

```
<vmm> ERROR:vmm/src/lib.rs:642 -- Error when adding new device to the VM: DeviceManager(MissingVirtualIommu)
```

To fix this issue, we will need to support creating `vIOMMU` device on-the-fly, instead of only allowing its creation at VM boot time based on user's initial CLI inputs.

**Comment by sboeuf** [comment]
>If we start a guest that does not require virtual IOMMU, we will hit the following error if we hotplug a device with iommu=on after VM boot

I believe this is even wider than that. If we try to hotplug a device with `iommu=on`, it shouldn't work.

>To fix this issue, we will need to support creating vIOMMU device on-the-fly

This isn't possible. An IOMMU is not a PCI device, you can't just hotplug it :-/

This issue is kind of a duplicate of #911.
There are two aspects here:
- We want to use a vIOMMU even when the VM hasn't been started with one because no devices are actually using it. Well, to solve this case, we must add an option to be able to create a vIOMMU even if no device is attached to it at boot.
- We want to be able to hotplug a device to this vIOMMU, well the only way is by allowing the b/d/f at the time the vIOMMU is created. This might means we want to enable a specific range of PCI devices to be attached, but I guess the best way would be to have a dedicated PCI bus to solve that issue. Given that @rbradford is working on supporting multiple PCI buses, this might be the right time to implement what's missing for virtio-iommu.

@eadamsintel could you describe your use case here as well, since it's always good to provide reasons for us to implement a feature.

**Comment by likebreath** [comment]
> I believe this is even wider than that. If we try to hotplug a device with iommu=on, it shouldn't work.

Great to know.

> This isn't possible. An IOMMU is not a PCI device, you can't just hotplug it :-/
> This issue is kind of a duplicate of #911.

Aha, right. Let's keep the issue for know until we make a decision about whether we want to add an option to create a vIOMMU at boot time (even no devices use it at the beginning). 

> could you describe your use case here as well, since it's always good to provide reasons for us to implement a feature.

I think one use case is Kata, where it always start a VM with minimum configuration and hotplug devices after the VM is booted.

**Comment by eadamsintel** [comment]
The use case for me is with regards to Intel® QuickAssist and its virtual functions for using with Kata Containers. The new process for using the upstream QAT driver is to bind its vf's to `vfio-pci` to work with QAT lib and openssl. This needs to happen for both the bare metal use case and virtualized use case. In the case of a VM, the vf's need to be bound to `vfio-pci` on the host and made available inside the container. Kata containers does this through hot plugging. Once inside the VM/container the QAT vf also needs to be bounded to `vfio-pci`. However, the QAT lib expects that each QAT vf is in its own IOMMU group.  When the QAT vf is cold booted into the VM it was put in its own IOMMU group. However hot plugging did not work. 

So for this to work for the QAT vf use case with Kata Containers I see the following possible options
1. QAT lib has to be updated to allow other non QAT devices to exist in the same IOMMU group. This could have security concerns and may not be acceptable to the QAT lib maintainers. 
2. Kata Containers needs to be updated to support cold booting of PCIe vf's. This might be tricky with some good reasons to not do this. Kata currently hot plugs everything. It can get tricky with the empty start containers to get this supported with Kata and could be problematic.
3. Cloud Hypervisor needs to be able to hot plug a PCIe vf into its own IOMMU group.

**Comment by sboeuf** [comment]
Yes I think the use case is totally valid and I agree Cloud Hypervisor should eventually support it.
I'll assess what's needed in order to support this, and try to come up with the best tradeoff, I'll keep you both posted :)

**Comment by sboeuf** [comment]
Yeah I'm not worried about adding a way to enable the vIOMMU even when no devices are attached to it, but it's more about which range of PCI devices do we make part of the vIOMMU. One option would be to attach any device that gets added later to it, but that means any PCI device being hotplugged will end up behind the vIOMMU no matter what.

**Comment by sboeuf** [comment]
@eadamsintel 
So you built the guest kernel with `CONFIG_VFIO_NOIOMMU`, but you also need to provide an extra parameter in the guest kernel command line `vfio.enable_unsafe_noiommu_mode=1`:
```
./cloud-hypervisor ... -cmdline "... vfio.enable_unsafe_noiommu_mode=1 ..." ...
```

Once your VM is booted, you can check the parameter is correctly enabled by doing:
```
cat /sys/module/vfio/parameters/enable_unsafe_noiommu_mode
```
which should return `Y`.

At this point, there's no IOMMU group in `/sys/kernel/iommu_groups`, that's expected.
The next step is to do what you described above:
- Unbind the device from its driver
- Bind it to `vfio-pci` using the `new_id` mechanism

At this point a new device should have appeared `/dev/vfio/noiommu-0`, and the corresponding IOMMU group 0 should have been created `/sys/kernel/iommu_groups/0`.

I think the guest kernel boot option `vfio.enable_unsafe_noiommu_mode=1` is what you were missing.
